examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py

```diff
@@ -1283,13 +1322,26 @@ def compute_embeddings(
                             alpha_schedule,
                             sigma_schedule,
                         )
+                        uncond_pred_noise = predicted_source_noise(
+                            uncond_teacher_output,
+                            start_timesteps,
+                            noisy_model_input,
+                            noise_scheduler.config.prediction_type,
+                            alpha_schedule,
+                            sigma_schedule,
+                        )
 
-                        # 20.4.11. Perform "CFG" to get x_prev estimate (using the LCM paper's CFG formulation)
+                        # 3. Calculate the CFG estimate of x_0 (pred_x0) and eps_0 (pred_noise)
+                        # Note that this uses the LCM paper's CFG formulation rather than the Imagen CFG formulation
                         pred_x0 = cond_pred_x0 + w * (cond_pred_x0 - uncond_pred_x0)
-                        pred_noise = cond_teacher_output + w * (cond_teacher_output - uncond_teacher_output)
+                        pred_noise = cond_pred_noise + w * (cond_pred_noise - uncond_pred_noise)
```

Comment:
Hmm this seems like a big change to me. I wonder how much of an effect it has on the results.

Reply:
Hi @sayakpaul, this change should be a no-op when the teacher LDM checkpoint has `self.scheduler.config.prediction_type == "epsilon"` (as is the case for the [`runwayml/stable-diffusion-v1-5`](https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/main/scheduler/scheduler_config.json) and [`stabilityai/stable-diffusion-xl-base-1.0`](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/scheduler/scheduler_config.json) checkpoints suggested in the `README` examples). The current logic of directly using the residual output of the teacher `unet` is correct when the `prediction_type` is `epsilon`:

https://github.com/huggingface/diffusers/blob/4836cfad9836e6742a1d09462f85313534388a48/examples/consistency_distillation/train_lcm_distill_lora_sdxl_wds.py#L1289

but in general we need to convert the residual `unet` output to a predicted epsilon for use in `DDIMSolver.ddim_step`, which is done by the new `predicted_source_noise` function, which is based on `DDIMScheduler.step`:

https://github.com/huggingface/diffusers/blob/4836cfad9836e6742a1d09462f85313534388a48/src/diffusers/schedulers/scheduling_ddim.py#L412-L422

Note that `pred_epsilon` is exactly equal to `model_output` when `prediction_type == "epsilon"`.

Reply:
Thanks for the solid explanation!